### PR TITLE
Feature no delay

### PR DESCRIPTION
**Pull request by cpehle**
This addresses #325, as far as I can tell there is no downside to this change. One thing that I noticed is that there are also some inconsistencies in how we handle resets and the naming conventions for current / spike integration. They should probably be adressed in a separate change.

### CODE REVIEW

File: norse/torch/functional/lif.py
Comment (Jegp) [inline norse/torch/functional/lif.py:432]: Why not use the cpp code? If you are concerned about parameterizing the surrogate gradient method, I suggest renaming the function to include `super` and then do the differentiation in the `LIF` module, such so we retain the performance benefits.

File: norse/torch/functional/lif.py
Comment (Jegp) [inline norse/torch/functional/lif.py:492]: Circular reference.

**Review by Jegp — CHANGES_REQUESTED** [review]
Thanks for the changes!

Before I can approve this, I still need to understand the motivation for this. Specifically, I would like some comparison that shows the need for this kind of change. Does it really provide the numerical stability we want? Does it actually solve the problem?

In principle, I think this can work, but this is also a big change in the fundamental workings of Norse. If we push this through, we need to make sure that all the tests are passing, notebooks are working, and the benchmarks are up to date.

File: norse/torch/functional/lif.py
Comment (cpehle) [reply norse/torch/functional/lif.py:492]: Yeah, I need to fix that.

File: norse/torch/functional/lif.py
Comment (cpehle) [reply norse/torch/functional/lif.py:432]: I don't think special casing like this is a good idea, right now. It breaks the expectation of the users and violates the interface.

Comment (Jegp) [reply norse/torch/functional/lif.py:432]: We already have those "special cases": https://github.com/norse/norse/blob/master/norse/torch/module/lif.py#L65

**Comment by Jegp** [comment]
> This is really not about numerical stability. It is about whether spikes get integrated (in discrete time) immediately by the following layer, or not. As was pointed out in the discussion, the way we do it right now has the undesirable effect of accumulating latency if many layers are sequentially connected. 

Unfortunately, this means that the FF implementations will be inconsistent with the recurrent ones. I can imagine that situation in the short run, but I don't think it's desirable in the long run. For me consistency >> "undesirable" delays in deep layers. 
One possible path forward would be to accept this as status quo, but then decide on a solid separation of synapse and neuron dynamics that is consistent across models and layer types.

File: norse/torch/functional/lif.py
Comment (cpehle) [reply norse/torch/functional/lif.py:432]: I wouldn't say this is a special case, just a different algorithm. The choice of surrogate gradient function at the moment does nothing *only* in the case of LIF neurons. This is unexpected and not consistent with the documentation.

Comment (Jegp) [reply norse/torch/functional/lif.py:432]: Overall, I think it's desirable to abstract the gradient algorithms in the "method" parameter. The error in LIF neurons, where it's ignored is critically bad. But that doesn't mean we can't still have it. Perhaps I'll finally try to flesh out the Threshold #242  module in the weekend. That could be a good place to start discussing how the API could/should look longer term.

Comment (cpehle) [reply norse/torch/functional/lif.py:432]: But it can't be abstracted like this, what is missing is proper integration support. Similarly for the threshold module, if you want to pass things to C++ you do not want to support necessarily arbitrary callable, which is why I went for the enum change. Baring that the proper way to do it would probably to just pass a callable and eliminate the threshold function altogether.

**Comment by cpehle** [comment]
I've now also tested that the mnist example notebook runs, with no degradation in performance. I think this change is good to go as is. I will expand the coverage back and we should try to integrate back the c++ code once it is sufficiently general.

**Review by Jegp — APPROVED** [review]
Despite the hit in performance after moving code from C++ -> Python, LGTM. Nice work :-)
And then I'll think about #89 / #54 so we can clearly measure where we are in terms of runtimes.

**Comment by LennardBo** [comment]
I don't mean to resurrect past discussion, but I was wondering why weren't the non-recurrent lif cells not included int his change? The [`_lif_feed_forward_step_jit`](https://github.com/norse/norse/blob/6df3bff7feba5cf2269c4e944625558c14d89a8f/norse/torch/functional/lif.py#L328) which to my understanding has the non-recurrent lif step still features the old "laggy" implementation.

Sorry that I didn't bring up this comment before the merge despite raising the sisue to begin with; during that time I was preoccupied with other projects and then was waiting for an official release of a newer version with these issues resolved (ofc that may take a while).

**Pull request merged**
